### PR TITLE
Cloud foundry specification

### PR DESCRIPTION
**Pull request by macsux**

### CODE REVIEW

File: source/Nuke.Common/Tools/CloudFoundry/CloudFoundryTasks.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+{
+    public static partial class CloudFoundryTasks
+    {
+        public static async Task CloudFoundryEnsureServiceReady(string serviceInstance)
+        {
+            bool IsCreating()
+            {    
+                var commandResult = CloudFoundryGetServiceInfo(c => c
+                        .SetServiceInstance(serviceInstance))
+                    .Where(x => x.Text.StartsWith("status:"))
```
Comment (robvanuden) [inline source/Nuke.Common/Tools/CloudFoundry/CloudFoundryTasks.cs]: Line 19 en 20 can be combined to `.First(x => x.Text.StartsWith("status:"))`

Comment (macsux) [reply source/Nuke.Common/Tools/CloudFoundry/CloudFoundryTasks.cs]: fixed

File: source/Nuke.Common/Tools/CloudFoundry/CloudFoundryTasks.cs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+            catch
+            {
+            }
+            
+            var cliDir = NukeBuild.TemporaryDirectory / "cf-cli";
+            var binary = cliDir / (Environment.OSVersion.Platform == PlatformID.Win32NT ? "cf.exe" : "cf");
+            if (FileExists(binary))
+                return binary;
+
+            string releaseName;
```
Comment (arodus) [inline source/Nuke.Common/Tools/CloudFoundry/CloudFoundryTasks.cs]: Personally I don't like that the executable is downloaded and installed automatically if not available. AFAIK we never do auto installing (besides dotnet). So I would prefer an exception and probably an install task.

What do you think about that @matkoch?

**Comment by macsux** [comment]
It's not really installed, it's locally mounted. Conceptually this is no
different then pulling a nupkg with exe inside that we do for many other
tools. Except that there is no official party to create and maintain those
nuget packages for cf cli.

**Comment by arodus** [comment]
The difference is that no NuGet package is added automatically, it must be always added manually to the build project so the user can decide on his own which version is used. With NuGet it's also possible to use different sources like local caches or company internal feeds (compliance requirements etc.).

With the current implementation it's also possible that the build suddenly stops working when the cli tool is updated and a breaking change was introduced.

**Comment by macsux** [comment]
Ok, I can build a side project to repackage official distros as nuget
packages. I'll refactor current PR to look for it in the environment and
add nuget based approach later as separate PR

**Comment by macsux** [comment]
I've created nuget packages for the CLI. As the CLI is multiplatform I split it up as multiple packages and used OS detection logic to pick the right one (RID based). It's expected that the consumer would add all relevant platforms they want to support in their build as nuget dependencies, and if they want to get fancy can use conditions inside csproj file to only include dependency base on the current OS type. Unfortunately, no easy way to automate this for the developer

**Comment by matkoch** [comment]
Are the CLIs too big? Because otherwise it makes more sense to have a single package containing all of them.

**Comment by macsux** [comment]
They are like 7 to 8mb each, I didn't want to make a single package that's
like 30mb and you only need one cli.

**Comment by matkoch** [comment]
Thanks for your contribution! 😄🎉

**Pull request merged**